### PR TITLE
When reopening a project no dynamic widgets available

### LINKED ISSUE #10595: When Reopening a Project, no dynamic widgets available.
The GUI attaches the correct visualizations but no response from the backend of the evaluated json.

Snag_17f6fc27

The simple workflow above works when being built but fails when reopened.

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #10595 

Changelog:
- fix: storing IdMap from the request
- update: apply updated external ids after loading the module from cache
- update: simplify the node invalidation logic

### Important Notes

https://github.com/user-attachments/assets/c93df161-3bc5-4ecf-8696-239a2498dfb8

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/Compiler.scala [hunk 1]
```diff
@@ -560,6 +564,21 @@ class Compiler(
...
+            context.getIr(module),
+            context.getIdMap(module)
+          )
+        if (updatedIr != null) {
+          context.updateModule(
+            module,
+            u => {
+              u.resetScope()
+              u.ir(updatedIr)
+              u.compilationStage(CompilationStage.AFTER_PARSING)
```
Comment (4e6) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/Compiler.scala:578]: We have to re-run the compiler passes if ids are updated

File: engine/runtime-compiler/src/main/java/org/enso/compiler/phase/IdMapProcessor.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  /**
+   * Applies the map with external ids to the module ir.
+   *
+   * @param ir the module ir.
+   * @param idMap the provided IdMap.
+   */
+  public static Module updateIr(Module ir, IdMap idMap) {
+    if (idMap == null) {
+      return null;
```
Comment (hubertp) [inline engine/runtime-compiler/src/main/java/org/enso/compiler/phase/IdMapProcessor.java:24]: Shouldn't this return `ir` itself and the usage site should check for null `idMap`?

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/PendingEdits.scala [hunk 3]
```diff
@@ -20,6 +21,20 @@ trait PendingEdits {
     */
   def dequeue(file: File): Seq[PendingEdit]
 
+  /** Put the IdMap of the file.
+    *
+    * @param file the edited file
+    * @param idMap the new IdMap
+    */
+  def putIdMap(file: File, idMap: IdMap): Unit
```
Comment (hubertp) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/PendingEdits.scala]: naming nit: `updateIdMap` or `setIdMap`?

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/PendingEdits.scala [hunk 4]
```diff
@@ -20,6 +21,20 @@ trait PendingEdits {
...
+
+  /** Consume the idMap of the file.
+    *
+    * @param file the edited file
+    * @return the IdMap of the file
+    */
+  def takeIdMap(file: File): Option[IdMap]
```
Comment (hubertp) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/execution/PendingEdits.scala]: `removeIdMap` or `getAndRemove`. `take` doesn't really say that we are removing the association.

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/UpsertVisualizationJob.scala [hunk 5]
```diff
@@ -618,13 +618,9 @@ object UpsertVisualizationJob {
             val externalId = expressionId
             module.getIr.preorder
               .find(_.getExternalId.contains(externalId))
-              .collect {
-                case name: Name.Literal =>
```
Comment (hubertp) [inline engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/UpsertVisualizationJob.scala:622]: why that change?

**Review by hubertp — APPROVED** [review]

File: engine/runtime-compiler/src/main/java/org/enso/compiler/phase/IdMapProcessor.java (hunk 2 above)
Comment (4e6) [reply engine/runtime-compiler/src/main/java/org/enso/compiler/phase/IdMapProcessor.java:24]: The non-null return value indicates that the `IR` was updated we need to re-run the compiler passes on it

File: engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/UpsertVisualizationJob.scala (hunk 5 above)
Comment (4e6) [reply engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/UpsertVisualizationJob.scala:622]: Nowadays it is enough to query the dataflow analysis. I assume the structure of dependencies graph has changed with time

File: engine/runtime-compiler/src/main/java/org/enso/compiler/phase/IdMapProcessor.java (hunk 2 above)
Comment (4e6) [reply engine/runtime-compiler/src/main/java/org/enso/compiler/phase/IdMapProcessor.java:24]: In most cases it is unchanged, and we can skip the compilation

**Review by JaroslavTulach — APPROVED** [review]

**Pull request merged**